File: internal/runctx/runctx.go
// Package runctx works out the run metadata an agent should never have to type.
//
// The key names are the canon vocabulary: OpenTelemetry's where OTel has a
// word, `krowk.`-namespaced where it does not. Nothing here writes a flat
// legacy key (`repo`, `commit`, …) — readers fall back to those, writers moved.
package runctx

import (
	"encoding/json"
	"os/exec"
	"regexp"
	"strings"
)

// Metadata is the JSON blob that rides along with a record. Flags override
// every detected field; omitempty does the pruning.
//
// One struct serves both records: a run carries all of it, an artifact carries
// the production snapshot only — Artifact() clears the run-only facts.
type Metadata struct {
	RepoName string `json:"vcs.repository.name,omitempty"`
	RepoURL  string `json:"vcs.repository.url.full,omitempty"`
	Commit   string `json:"vcs.ref.head.revision,omitempty"`
	Branch   string `json:"vcs.ref.head.name,omitempty"`
	// Dirty is a pointer so "clean tree" is recorded as false while "not a git
	// checkout" records nothing at all — every key is optional, never defaulted.
	Dirty   *bool  `json:"krowk.vcs.dirty,omitempty"`
	Harness string `json:"krowk.harness,omitempty"`
	// Model and System are recorded only when the environment actually says.
	// KROWK_MODEL is the harness-agnostic spelling any wrapper can export;
	// the vendor variables are the signals specific harnesses already honour.
	// No signal, no key: a guessed model in an audit record is worse than
	// none, and --metadata can always spell one out.
	Model  string `json:"gen_ai.request.model,omitempty"`
	System string `json:"gen_ai.system,omitempty"`
	Client string `json:"krowk.client,omitempty"`

	// Facts about the work, not about one file: run-only.
	ChangeID    string   `json:"vcs.change.id,omitempty"`
	ChangeTitle string   `json:"vcs.change.title,omitempty"`
	ChangeURL   string   `json:"krowk.change.url,omitempty"`
	Session     string   `json:"krowk.session,omitempty"`
	References  []string `json:"krowk.references,omitempty"`

	// remoteSlug remembers what the origin remote itself named, so an
	// overridden repository name can be checked against it. Never serialized.
	remoteSlug string
}

// Env is a lookup function, so tests do not have to touch the process
// environment. Pass os.Getenv.
type Env func(string) string

// Overrides are the metadata fields a caller supplies rather than detects.
// Empty strings leave the detected value alone; the fields that cannot be
// detected at all are simply carried through.
type Overrides struct {
	Repo        string
	Commit      string
	Agent       string
	PullRequest string
	References  []string
	Session     string
	Title       string
	Client      string
}

// Resolve detects what it can and lets the caller's values win. Shared by the
// CLI and the MCP server so both report metadata the same way.
func Resolve(env Env, o Overrides) Metadata {
	m := Detect(env)
	override(&m.RepoName, o.Repo)
	override(&m.Commit, o.Commit)
	override(&m.Harness, o.Agent)
	override(&m.ChangeURL, o.PullRequest)
	override(&m.Session, o.Session)
	m.References = o.References
	m.ChangeTitle = o.Title
	m.Client = o.Client
	// An overridden harness re-decides the provider: the claim has to follow
	// the correction, not the detection it replaced.
	m.System = DetectSystem(m.Harness, m.Model)
	m.ChangeID = ChangeID(m.ChangeURL)
	m.reconcileRepo()
	return m
}

func override(dst *string, v string) {
	if v != "" {
		*dst = v
	}
}

// Detect fills in everything that can be read off the machine.
func Detect(env Env) Metadata {
	remote := git("remote", "get-url", "origin")

	m := Metadata{
		RepoName:   firstNonEmpty(env("GITHUB_REPOSITORY"), Slug(remote)),
		RepoURL:    RepoURL(env, remote),
		Commit:     firstNonEmpty(env("GITHUB_SHA"), git("rev-parse", "HEAD")),
		Branch:     Branch(env),
		Dirty:      dirty(),
		Harness:    DetectAgent(env),
		Model:      DetectModel(env),
		Session:    DetectSession(env),
		ChangeURL:  CIPullRequest(env),
		remoteSlug: Slug(remote),
	}
	m.System = DetectSystem(m.Harness, m.Model)
	m.ChangeID = ChangeID(m.ChangeURL)
	m.reconcileRepo()
	return m
}

// Branch names the branch the work is on. CI checks out a detached HEAD, so
// `git rev-parse --abbrev-ref HEAD` answers the literal string "HEAD" there —
// the branch is in the environment instead: GITHUB_HEAD_REF on a pull request
// (the source branch, where GITHUB_REF_NAME would say `412/merge`), and
// GITHUB_REF_NAME on a push. A local detached HEAD records no branch at all,
// which is the truth of it.
func Branch(env Env) string {
	if b := firstNonEmpty(env("GITHUB_HEAD_REF"), env("GITHUB_REF_NAME")); b != "" {
		return b
	}
	if b := git("rev-parse", "--abbrev-ref", "HEAD"); b != "HEAD" {
		return b
	}
	return ""
}

// DetectModel names the model that did the work. KROWK_MODEL first, because it
// is ours and any harness can export it; ANTHROPIC_MODEL is a signal Claude
// Code already honours, so when it is set it names the model in use.
func DetectModel(env Env) string {
	return firstNonEmpty(
		env("KROWK_MODEL"),
		env("ANTHROPIC_MODEL"),
	)
}

// DetectSystem names the provider, and only when something actually implies
// it: the model's own family, or Claude Code as the harness — that harness
// runs Anthropic models wherever they are hosted. Anything else stays unsaid
// rather than guessed; --metadata gen_ai.system=... spells out the rest.
func DetectSystem(harness, model string) string {
	switch {
	case strings.HasPrefix(model, "claude"):
		return "anthropic"
	case strings.HasPrefix(model, "gpt"):
		return "openai"
	case strings.HasPrefix(model, "gemini"):
		return "gcp.gemini"
	case harness == "claude-code":
		return "anthropic"
	}
	return ""
}

// Artifact is the production stamp for one file: the snapshot of this moment,
// minus the facts that belong to the work rather than the file — the change,
// the session, the references.
func (m Metadata) Artifact() Metadata {
	m.ChangeID, m.ChangeTitle, m.ChangeURL, m.Session = "", "", "", ""
	m.References = nil
	return m
}

// WithExtras lays caller-supplied keys over the stamp. The caller's value wins
// on any collision, the standard keys included: those are detected or
// defaulted, so a caller spelling one out is correcting a detection.
func (m Metadata) WithExtras(extras map[string]string) any {
	if len(extras) == 0 {
		return m
	}
	b, _ := json.Marshal(m)
	out := map[string]any{}
	_ = json.Unmarshal(b, &out)
	for k, v := range extras {
		out[k] = v
	}
	return out
}

// reconcileRepo drops the repository URL when the name and the URL disagree:
// the URL is read off the checkout's remote while the name can arrive by flag
// or from CI, and a link to the wrong repository is worse than no link.
func (m *Metadata) reconcileRepo() {
	if m.remoteSlug != "" && m.RepoName != m.remoteSlug {
		m.RepoURL = ""
	}
}

// RepoURL is the repository's home on its host. An https remote names it
// directly; an ssh remote gets one built only for a host whose web shape is
// known, rather than guessing a path that 404s.
func RepoURL(env Env, remote string) string {
	if strings.HasPrefix(remote, "http://") || strings.HasPrefix(remote, "https://") {
		return strings.TrimSuffix(strings.TrimRight(remote, "/"), ".git")
	}
	slug := Slug(remote)
	if slug == "" {
		return ""
	}
	server := env("GITHUB_SERVER_URL") // set by Actions, and correct on Enterprise
	if server == "" && Host(remote) == "github.com" {
		server = "https://github.com"
	}
	if server == "" {
		return ""
	}
	return strings.TrimRight(server, "/") + "/" + slug
}

// DetectSession finds the agent run this upload belongs to, so --session is a
// correction rather than something the agent has to remember.
func DetectSession(env Env) string {
	return firstNonEmpty(
		env("KROWK_SESSION"),
		env("CLAUDE_CODE_SESSION_ID"),
		env("CURSOR_TRACE_ID"),
		env("GITHUB_RUN_ID"),
	)
}

// DetectAgent names the harness driving the upload.
func DetectAgent(env Env) string {
	switch {
	case env("KROWK_AGENT") != "":
		return env("KROWK_AGENT")
	case env("CLAUDECODE") != "" || env("CLAUDE_CODE") != "":
		return "claude-code"
	case env("CURSOR_TRACE_ID") != "":
		return "cursor"
	case env("GITHUB_ACTIONS") != "":
		return "github-actions"
	}
	return ""
}

var (
	slugRE     = regexp.MustCompile(`[:/]([^/:]+/[^/]+?)(?:\.git)?/?$`)
	prRefRE    = regexp.MustCompile(`^refs/pull/(\d+)/`)
	hostRE     = regexp.MustCompile(`^(?:[a-z+]+://)?(?:[^@/]+@)?([^/:]+)[:/]`)
	changeIDRE = regexp.MustCompile(`/(\d+)/?$`)
)

// Host pulls github.com out of either remote spelling.
func Host(remote string) string {
	if m := hostRE.FindStringSubmatch(remote); m != nil {
		return m[1]
	}
	return ""
}

// Slug turns git@github.com:acme/storefront.git into acme/storefront.
func Slug(remote string) string {
	if m := slugRE.FindStringSubmatch(remote); m != nil {
		return m[1]
	}
	return ""
}

// ChangeID is the change's number, derived from its URL — OTel has a key for
// the number but none for the URL, so the URL is what gets recorded raw.
func ChangeID(changeURL string) string {
	if m := changeIDRE.FindStringSubmatch(changeURL); m != nil {
		return m[1]
	}
	return ""
}

// CIPullRequest turns refs/pull/412/merge into the pull request URL.
func CIPullRequest(env Env) string {
	m := prRefRE.FindStringSubmatch(env("GITHUB_REF"))
	repo := env("GITHUB_REPOSITORY")
	if m == nil || repo == "" {
		return ""
	}
	return "https://github.com/" + repo + "/pull/" + m[1]
}

// dirty reports the worktree state, or nil outside a git checkout — the
// distinction a plain bool cannot carry.
func dirty() *bool {
	out, err := exec.Command("git", "status", "--porcelain").Output()
	if err != nil {
		return nil
	}
	d := strings.TrimSpace(string(out)) != ""
	return &d
}

// git returns trimmed stdout, or "" for any failure — no git, no repo, no remote.
func git(args ...string) string {
	out, err := exec.Command("git", args...).Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func firstNonEmpty(values ...string) string {
	for _, v := range values {
		if v != "" {
			return v
		}
	}
	return ""
}

File: crates/krowk/src/runctx.rs
//! Where the work came from: the repository, commit, branch and dirty flag,
//! the agent and its session, the pull request — detected from git and the
//! environment, and overridden by whatever the caller said.

use serde::Serialize;
use std::process::Command;

/// The process environment, as the CLI reads it.
pub type Env<'a> = &'a dyn Fn(&str) -> String;

/// The run context, keyed by the OpenTelemetry names canon adopts. Empty
/// fields are left out: an absent key and an empty one mean the same, and
/// the absent one is shorter on every card.
#[derive(Debug, Clone, Default, Serialize, PartialEq)]
pub struct Metadata {
    #[serde(rename = "vcs.repository.name", skip_serializing_if = "String::is_empty")]
    pub repo_name: String,
    #[serde(rename = "vcs.repository.url.full", skip_serializing_if = "String::is_empty")]
    pub repo_url: String,
    #[serde(rename = "vcs.ref.head.revision", skip_serializing_if = "String::is_empty")]
    pub commit: String,
    #[serde(rename = "vcs.ref.head.name", skip_serializing_if = "String::is_empty")]
    pub branch: String,
    /// None outside a git checkout — the distinction a plain bool cannot carry.
    #[serde(rename = "krowk.vcs.dirty", skip_serializing_if = "Option::is_none")]
    pub dirty: Option<bool>,
    #[serde(rename = "krowk.harness", skip_serializing_if = "String::is_empty")]
    pub harness: String,
    #[serde(rename = "gen_ai.request.model", skip_serializing_if = "String::is_empty")]
    pub model: String,
    #[serde(rename = "gen_ai.system", skip_serializing_if = "String::is_empty")]
    pub system: String,
    #[serde(rename = "krowk.client", skip_serializing_if = "String::is_empty")]
    pub client: String,
    #[serde(rename = "vcs.change.id", skip_serializing_if = "String::is_empty")]
    pub change_id: String,
    #[serde(rename = "vcs.change.title", skip_serializing_if = "String::is_empty")]
    pub change_title: String,
    #[serde(rename = "krowk.change.url", skip_serializing_if = "String::is_empty")]
    pub change_url: String,
    #[serde(rename = "krowk.session", skip_serializing_if = "String::is_empty")]
    pub session: String,
    #[serde(rename = "krowk.links", skip_serializing_if = "Vec::is_empty")]
    pub links: Vec<Link>,
    #[serde(rename = "krowk.references", skip_serializing_if = "Vec::is_empty")]
    pub references: Vec<String>,
    /// The slug the origin remote names, kept to judge whether an overridden
    /// repository still matches the URL detected for it.
    #[serde(skip)]
    remote_slug: String,
}

/// A link the work is about: the issue, the spec, the discussion.
#[derive(Debug, Clone, Default, Serialize, PartialEq)]
pub struct Link {
    pub url: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub title: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub rel: String,
}

pub const MAX_LINKS: usize = 20;
pub const MAX_LINK_URL: usize = 2048;
pub const MAX_LINK_TITLE: usize = 140;
pub const MAX_LINK_REL: usize = 64;
pub const MAX_LINKS_BYTES: usize = 8192;

/// The rels a reader recognises; any other word is recorded as given.
pub const LINK_RELS: &[&str] = &["tracks", "fixes", "spec", "discussion", "source", "supersedes"];

/// Refuses links a card could not show honestly: too many, too long, not an
/// absolute http(s) URL, carrying credentials, or text that repaints a row.
pub fn validate_links(links: &[Link]) -> Result<(), String> {
    if links.len() > MAX_LINKS {
        return Err(format!(
            "{} links is more than the {MAX_LINKS} a run records: link what the work is about, not everything it touched",
            links.len()
        ));
    }
    for (i, l) in links.iter().enumerate() {
        let at = format!("link {}", i + 1);
        validate_link_url(&at, &l.url)?;
        validate_link_line(&at, "title", &l.title, MAX_LINK_TITLE)?;
        validate_link_line(&at, "rel", &l.rel, MAX_LINK_REL)?;
    }
    let bytes = if links.is_empty() { 0 } else { serde_json::to_string(links).map_or(0, |s| s.len()) };
    if bytes > MAX_LINKS_BYTES {
        return Err(format!(
            "the links come to {bytes} bytes of metadata, past the {MAX_LINKS_BYTES} they may fill: a run's metadata budget is shared with everything krowk detects"
        ));
    }
    Ok(())
}

fn validate_link_line(at: &str, field: &str, value: &str, max: usize) -> Result<(), String> {
    let n = value.chars().count();
    if n > max {
        return Err(format!("{at} has a {n}-character {field}, past the {max} one line holds"));
    }
    if value.chars().any(display_hostile) {
        return Err(format!(
            "{at} has a {field} that would repaint the row it is drawn on — a line break, a control character or a bidi override: it is what a reader sees instead of the URL, on one row"
        ));
    }
    Ok(())
}

/// What a single row must not carry: controls, line and paragraph
/// separators, and the invisible characters that reorder text.
fn display_hostile(c: char) -> bool {
    c != '\u{200d}' && (c.is_control() || matches!(c, '\u{2028}' | '\u{2029}' | '\u{200b}'..='\u{200f}') || crate::termclean::reordering(c))
}

fn validate_link_url(at: &str, raw: &str) -> Result<(), String> {
    if raw.trim().is_empty() {
        return Err(format!("{at} has no url: every link needs an absolute http(s) URL"));
    }
    if raw.chars().any(|c| c == ' ' || display_hostile(c)) {
        return Err(format!(
            "{at} ({raw:?}) has a space, a control character or a bidi override in it: a URL is one unbroken word, and one that reads as it resolves"
        ));
    }
    let n = raw.chars().count();
    if n > MAX_LINK_URL {
        return Err(format!("{at} is {n} characters, past the {MAX_LINK_URL} a URL may be"));
    }
    let not_url = || format!("{at} ({raw}) is not an absolute http(s) URL — a ticket key or an internal ID belongs in references, not in links");
    let (scheme, rest) = raw.split_once("://").ok_or_else(not_url)?;
    if !scheme.eq_ignore_ascii_case("http") && !scheme.eq_ignore_ascii_case("https") {
        return Err(not_url());
    }
    let end = rest.find(['/', '?', '#']).unwrap_or(rest.len());
    let (authority, tail) = rest.split_at(end);
    let host = authority.rsplit_once('@').map_or(authority, |(_, h)| h);
    if host.is_empty() {
        return Err(not_url());
    }
    if authority.contains('@') {
        return Err(format!(
            "{at} carries a username or password in the URL, and run metadata is public and permanent: link {scheme}://{host}{tail} instead"
        ));
    }
    Ok(())
}

/// What the caller said, which wins over what was detected.
#[derive(Debug, Clone, Default)]
pub struct Overrides {
    pub repo: String,
    pub commit: String,
    pub agent: String,
    pub pull_request: String,
    pub links: Vec<Link>,
    pub references: Vec<String>,
    pub session: String,
    pub title: String,
    pub client: String,
}

/// Detection with the caller's overrides applied.
pub fn resolve(env: Env, o: Overrides) -> Metadata {
    let mut m = detect(env);
    let over = |dst: &mut String, v: String| {
        if !v.is_empty() {
            *dst = v;
        }
    };
    over(&mut m.repo_name, o.repo);
    over(&mut m.commit, o.commit);
    over(&mut m.harness, o.agent);
    over(&mut m.change_url, o.pull_request);
    over(&mut m.session, o.session);
    m.links = o.links;
    m.references = o.references;
    m.change_title = o.title;
    m.client = o.client;
    m.system = detect_system(&m.harness, &m.model).into();
    m.change_id = change_id(&m.change_url);
    m.reconcile_repo();
    m
}

/// What git and the environment say, with nothing overridden.
pub fn detect(env: Env) -> Metadata {
    let remote = git(&["remote", "get-url", "origin"]);
    let mut m = Metadata {
        repo_name: first(&[env("GITHUB_REPOSITORY"), slug(&remote)]),
        repo_url: repo_url(env, &remote),
        commit: first(&[env("GITHUB_SHA"), git(&["rev-parse", "HEAD"])]),
        branch: branch(env),
        dirty: dirty(),
        harness: detect_agent(env),
        model: first(&[env("KROWK_MODEL"), env("ANTHROPIC_MODEL")]),
        session: first(&[env("KROWK_SESSION"), env("CLAUDE_CODE_SESSION_ID"), env("CURSOR_TRACE_ID"), env("GITHUB_RUN_ID")]),
        change_url: ci_pull_request(env),
        remote_slug: slug(&remote),
        ..Metadata::default()
    };
    m.system = detect_system(&m.harness, &m.model).into();
    m.change_id = change_id(&m.change_url);
    m.reconcile_repo();
    m
}

impl Metadata {
    /// The production record stamped on each artifact: the work-level facts
    /// — the change, the session, the links — live on the run.
    pub fn artifact(&self) -> Metadata {
        Metadata {
            change_id: String::new(),
            change_title: String::new(),
            change_url: String::new(),
            session: String::new(),
            links: Vec::new(),
            references: Vec::new(),
            ..self.clone()
        }
    }

    /// The record with the caller's --metadata pairs on top; theirs wins.
    pub fn with_extras(&self, extras: &[(String, String)]) -> serde_json::Value {
        let mut v = serde_json::to_value(self).expect("metadata serializes");
        if let serde_json::Value::Object(map) = &mut v {
            for (k, val) in extras {
                map.insert(k.clone(), serde_json::Value::String(val.clone()));
            }
        }
        v
    }

    /// A repository named by override is not the one the detected URL points
    /// at, so the URL goes rather than contradict it.
    fn reconcile_repo(&mut self) {
        if !self.remote_slug.is_empty() && self.repo_name != self.remote_slug {
            self.repo_url.clear();
        }
    }
}

fn first(values: &[String]) -> String {
    values.iter().find(|v| !v.is_empty()).cloned().unwrap_or_default()
}

fn git(args: &[&str]) -> String {
    Command::new("git")
        .args(args)
        .output()
        .ok()
        .filter(|o| o.status.success())
        .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
        .unwrap_or_default()
}

/// The branch: CI's word first, then git's, and none on a detached head.
pub fn branch(env: Env) -> String {
    let head_ref = env("GITHUB_HEAD_REF");
    if !head_ref.is_empty() {
        return head_ref;
    }
    let ref_name = env("GITHUB_REF_NAME");
    if !ref_name.is_empty() && env("GITHUB_REF_TYPE") == "branch" {
        return ref_name;
    }
    let b = git(&["rev-parse", "--abbrev-ref", "HEAD"]);
    if b == "HEAD" { String::new() } else { b }
}

/// The model vendor, from the model's name or, failing that, the harness.
pub fn detect_system(harness: &str, model: &str) -> &'static str {
    if model.starts_with("claude") {
        "anthropic"
    } else if model.starts_with("gpt") {
        "openai"
    } else if model.starts_with("gemini") {
        "gcp.gemini"
    } else if harness == "claude-code" {
        "anthropic"
    } else {
        ""
    }
}

/// The harness driving the upload.
pub fn detect_agent(env: Env) -> String {
    let agent = env("KROWK_AGENT");
    if !agent.is_empty() {
        agent
    } else if !env("CLAUDECODE").is_empty() || !env("CLAUDE_CODE").is_empty() {
        "claude-code".into()
    } else if !env("CURSOR_TRACE_ID").is_empty() {
        "cursor".into()
    } else if !env("GITHUB_ACTIONS").is_empty() {
        "github-actions".into()
    } else {
        String::new()
    }
}

/// The web URL of the repository. An https remote names it directly, with
/// any credentials in it dropped — CI clones as https://x-access-token:<token>@…,
/// and this lands in public metadata. An ssh remote gets one only where the
/// host's web shape is known.
pub fn repo_url(env: Env, remote: &str) -> String {
    if remote.starts_with("http://") || remote.starts_with("https://") {
        let clean = without_userinfo(remote);
        let clean = clean.trim_end_matches('/');
        return clean.strip_suffix(".git").unwrap_or(clean).to_string();
    }
    let slug = slug(remote);
    if slug.is_empty() {
        return String::new();
    }
    let mut server = env("GITHUB_SERVER_URL");
    if server.is_empty() && host(remote) == "github.com" {
        server = "https://github.com".into();
    }
    if server.is_empty() {
        return String::new();
    }
    format!("{}/{slug}", server.trim_end_matches('/'))
}

fn without_userinfo(remote: &str) -> String {
    let (scheme, rest) = remote.split_once("://").unwrap_or((remote, ""));
    let (authority, path) = match rest.split_once('/') {
        Some((a, p)) => (a, Some(p)),
        None => (rest, None),
    };
    let authority = authority.rsplit_once('@').map_or(authority, |(_, h)| h);
    match path {
        Some(p) => format!("{scheme}://{authority}/{p}"),
        None => format!("{scheme}://{authority}"),
    }
}

/// `owner/name` out of any remote spelling.
pub fn slug(remote: &str) -> String {
    static RE: std::sync::LazyLock<regex_lite::Regex> =
        std::sync::LazyLock::new(|| regex_lite::Regex::new(r"[:/]([^/:]+/[^/]+?)(?:\.git)?/?$").unwrap());
    RE.captures(remote).map(|c| c[1].to_string()).unwrap_or_default()
}

fn host(remote: &str) -> String {
    static RE: std::sync::LazyLock<regex_lite::Regex> =
        std::sync::LazyLock::new(|| regex_lite::Regex::new(r"^(?:[a-z+]+://)?(?:[^@/]+@)?([^/:]+)[:/]").unwrap());
    RE.captures(remote).map(|c| c[1].to_string()).unwrap_or_default()
}

/// The change's number, from its URL.
pub fn change_id(url: &str) -> String {
    static RE: std::sync::LazyLock<regex_lite::Regex> =
        std::sync::LazyLock::new(|| regex_lite::Regex::new(r"/(\d+)/?$").unwrap());
    RE.captures(url).map(|c| c[1].to_string()).unwrap_or_default()
}

/// refs/pull/412/merge as the pull request's URL.
pub fn ci_pull_request(env: Env) -> String {
    static RE: std::sync::LazyLock<regex_lite::Regex> =
        std::sync::LazyLock::new(|| regex_lite::Regex::new(r"^refs/pull/(\d+)/").unwrap());
    let repo = env("GITHUB_REPOSITORY");
    match RE.captures(&env("GITHUB_REF")) {
        Some(c) if !repo.is_empty() => format!("https://github.com/{repo}/pull/{}", &c[1]),
        _ => String::new(),
    }
}

fn dirty() -> Option<bool> {
    let out = Command::new("git").args(["status", "--porcelain"]).output().ok().filter(|o| o.status.success())?;
    Some(!String::from_utf8_lossy(&out.stdout).trim().is_empty())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn env(pairs: &[(&str, &str)]) -> impl Fn(&str) -> String {
        let pairs: Vec<(String, String)> = pairs.iter().map(|(k, v)| (k.to_string(), v.to_string())).collect();
        move |k| pairs.iter().find(|(key, _)| key == k).map(|(_, v)| v.clone()).unwrap_or_default()
    }

    #[test]
    fn repo_url_names_the_web_page_and_never_the_credentials() {
        let none = env(&[]);
        for (remote, want) in [
            ("https://github.com/acme/storefront.git", "https://github.com/acme/storefront"),
            ("https://git.acme.dev/acme/storefront/", "https://git.acme.dev/acme/storefront"),
            ("git@github.com:acme/storefront.git", "https://github.com/acme/storefront"),
            ("git@gitlab.com:acme/storefront.git", ""),
            ("", ""),
            ("https://x-access-token:ghs_secret@github.com/acme/storefront.git", "https://github.com/acme/storefront"),
            ("https://user@git.acme.dev/acme/storefront", "https://git.acme.dev/acme/storefront"),
            ("https://a:b@c@github.com", "https://github.com"),
        ] {
            assert_eq!(repo_url(&none, remote), want, "{remote}");
        }
        let ghes = env(&[("GITHUB_SERVER_URL", "https://github.acme.dev/")]);
        assert_eq!(repo_url(&ghes, "git@github.acme.dev:acme/storefront.git"), "https://github.acme.dev/acme/storefront");
    }

    #[test]
    fn change_ids_and_ci_pull_requests_come_from_their_urls() {
        assert_eq!(change_id("https://github.com/acme/storefront/pull/412"), "412");
        assert_eq!(change_id("https://gitlab.com/acme/storefront/-/merge_requests/7/"), "7");
        assert_eq!(change_id("https://github.com/acme/storefront"), "");
        let ci = env(&[("GITHUB_REF", "refs/pull/42/merge"), ("GITHUB_REPOSITORY", "o/r")]);
        assert_eq!(ci_pull_request(&ci), "https://github.com/o/r/pull/42");
    }

    #[test]
    fn links_are_refused_for_what_a_card_could_not_show() {
        let link = |url: &str| Link { url: url.into(), ..Link::default() };
        assert!(validate_links(&[link("https://example.com/x")]).is_ok());
        for bad in ["ftp://x.example", "https://a b.example", "JIRA-1", "https://", "https://u:p@example.com/x"] {
            assert!(validate_links(&[link(bad)]).is_err(), "{bad}");
        }
        assert!(validate_links(&[Link { url: "https://e.x".into(), title: "a\nb".into(), rel: String::new() }]).is_err());
        assert!(validate_links(&vec![link("https://e.x"); 21]).is_err());
    }
}
